pages/blog/_posts/accidentally-quadratic-evaluating-trillions-of-event-matches-in-real-time.mdx

```diff
@@ -0,0 +1,166 @@
+
+This is a massive issue that kills scaling. At Inngest, we've seen timeouts, CPU spikes, and huge pressure in our data stores, which store and serve expressions. This isn't just mission-critical for customers; it is mission-critical for Inngest. How do we evaluate billions of expressions in real-time?
+
+Do the least amount of work possible.
+
+## Ignore Everything (Except The One Thing You Care About)
+
+The only insight that matters is this: A series of expressions must be matched, and one event must match against all expressions. We just need to know and not evaluate the expressions that don't match. If we can cut out the work, then we solve our problem.
+
+To do this, we need to do two things:
+
+1. Normalize our expressions and parse them into an abstract syntax tree, then
+2. Use hashmaps to look up equalities and find the exact match.
+
+### Normalizing Expressions and Building ASTs
+
+An abstract syntax tree (AST) is a hierarchical representation of the code's syntactic structure, where each node corresponds to a variable, expression, or statement.
+
+Using ASTs means we can break down each expression into constituent parts and then traverse only the nodes on the tree that match at each level. We must normalize the expressions before adding them to the tree, as this allows us to cache expression parsing into the tree.
+
+To do this, we lift variables from expressions. For example:
+
+```js
+event.data.id == "foo"
+```
+
+and:
+
+```js
+event.data.id == "bar"
+```
+
+both become:
+
+```js
+event.data.id == vars.a
+```
+
+where `vars.a` points to a different value (either “foo” or “bar”). We then parse the expressions into ASTs and normalize the AST so that differences in the CEL added by customers are …. For instance, `!(a == b)` is the same as `(a != b)`, and `!(a > b)` is the same as `(a <= b)`.
+
+![Diagram showing a conditional check and its representation in a structured format. The condition 'event.data.order_id == abc_123' is translated to a structured format with operator: 'equals', variable_ident: 'event.data.order_id', and literal_value: 'abc_123'. The diagram connects these two representations with arrows, indicating the relationship between the original condition and its structured format.](/assets/blog/accidentally-quadratic/graph-2-b&w.png)
+
+With each expression broken down, parsed, and normalized, we need to do the actual match.
+
+## Hashmaps & B-Trees FTW
+
+Initially, we tried some fancy ways to match equalities, such as aggregate trees, but we found simple hashmaps work best for string equality, while B-trees work best for numbers (we're exploring roaring bitmaps on the hash front, too).
+
+Our matching engine is a simple yet effective lookup for string equality matching within expressions. Here's how it works its magic:
```

Comment:
@djfarrelly: unresolved comment here in the Docs

Reply:
```suggestion
Our matching engine is a simple yet effective lookup for string equality matching within expressions. It's all written in [our open source `expr` repository](https://github.com/inngest/expr) in Go. Here's how it works its magic:
```